### ISSUE TITLE
preemptive retry for retrieve requests

### ISSUE THREAD

**Issue by acud** [ethersphere/bee#1082]
## Background
Retrieve request timeouts are discrete in nature, meaning, one peer must time-out before we try the next. This potentially means that a certain retrieve request cascade might take a very long time due to a slow peer, resulting in a general slow-down of the whole request cascade.

## Task
Try to request from the next peer _before_ current peer completely times out, but only after a certain amount of time has passed.
Example:
If I request a chunk from a peer and the timeout is `5s` long, I would like to ask for the chunk from another peer after 1 second, however I would like to allow the first peer to return the chunk while the preemptive retry is already in-flight. Once the chunk is retrieved from one of the peers, the other requests should be cancelled (i.e. libp2p streams should be `Reset()`).
In order for this to be compatible with #1081, the amount of time that should pass before a preemptive retrieval occurs should be defined as a function of the overall timeout given for the specific retrieve request, rather than a constant amount of time. On very short timeouts we might not want to do so at all.
It is possible that several retrieve requests will be in-flight simultaneously (i.e. more than 2). We should consider whether setting a hard limit for in-flight requests is necessary.

## Acceptance criteria
A preemptive request should happen when we request from a slow peer.

**Comment by Eknir** [comment]
We should note that cancelling a request should not come without a cost, as the peer is already doing work for you. As in a real contract, I would suggest that preemptively cancelling does incur some cost to the requester.

**Comment by Eknir** [comment]
Hi all,
Could you please let me know how the bandwidth incentives play out with this issue? When I request from more than one peer at the same time, will I also pay both of them if they give it to me?

**Comment by Eknir** [comment]
`Once the chunk is retrieved from one of the peers, the other requests should be cancelled (i.e. libp2p streams should be Reset()).`

If you are often requesting things from me and you cancel your order before I can deliver (but I have already asked my downstream peers), I don't want to serve you requests anymore. Canceling the request while it is ongoing in is a way to spam the network for free that could cause havoc in the network. 

@acud , could you please speak about this with @zelig (or if you have done this already, let me know the conclusions?)

**Comment by zbiljic** [comment]
Current implementation is described in PR for this issue, #1096.

**Comment by Eknir** [comment]
> `Current code will wait for any started retrieve request to complete, in order to credit any peer with successful delivery.`

Meaning you pay any peer who delivers to you, no matter whether you have already received it before or not? This seems better to me. However, the question still persists: what happens when another peer frequently cancels the request before we can deliver. This is, however, a separate issue.

**Comment by zbiljic** [comment]
> Meaning you pay any peer who delivers to you, no matter whether you have already received it before or not?

Yes, every peer contacted that successfully delivers before timeout will be credited.
This was done with taking into account the comment you made here.

> Nevertheless, the question still persists: what happens when another peer frequently cancels the request before we can deliver.

The current system does not have any concrete notion of cancelation. The peer is given request for chunk, and only "requester" knows how much time it has appointed (allowed) for peer to respond with the chunk. If the process of transferring chunk succeeds, both sides should account for transfer price.

I do not see anything protecting peer that attempts to respond with chunk, and connection closes before completing request fully.
